datashader/geo.py

```diff
@@ -0,0 +1,472 @@
+import datashader.transfer_functions as tf
+
+from datashader.colors import rgb
+from datashader.utils import ngjit
+from xarray import DataArray
+
+__all__ = ['mean', 'binary', 'slope', 'aspect', 'ndvi', 'hillshade', 'generate_terrain']
+
+
+def hillshade(agg, azimuth=225, angle_altitude=25):
+    """Illuminates 2D DataArray from specific azimuth and altitude.
+
+    Parameters
+    ----------
+    agg : DataArray
+    altitude : int, optional (default: 30)
+        Altitude angle of the sun specified in degrees.
+    azimuth : int, optional (default: 315)
+        The angle between the north vector and the perpendicular projection
+        of the light source down onto the horizon specified in degrees.
+    cmap : list of colors or matplotlib.colors.Colormap, optional
+        The colormap to use. Can be either a list of colors (in any of the
+        formats described above), or a matplotlib colormap object.
+        Default is `["lightgray", "black"]`
+    how : str or callable, optional
+        The hillshade method to use. Valid strings are 'mdow' [default],
+        'simple'.
+    alpha : int, optional
+        Value between 0 - 255 representing the alpha value of pixels which contain
+        data (i.e. non-nan values). Regardless of this value, `NaN` values are
+        set to fully transparent.
+
+    Returns
+    -------
+    Datashader Image
+
+    """
+    azimuth = 360.0 - azimuth
+    x, y = np.gradient(agg.data)
+    slope = np.pi/2. - np.arctan(np.sqrt(x*x + y*y))
+    aspect = np.arctan2(-x, y)
+    azimuthrad = azimuth*np.pi/180.
+    altituderad = angle_altitude*np.pi/180.
+    shaded = np.sin(altituderad) * np.sin(slope) + np.cos(altituderad) * np.cos(slope)*np.cos((azimuthrad - np.pi/2.) - aspect)
+    data = (shaded + 1) / 2
+    return DataArray(data, attrs=agg.attrs)
+
+
+@ngjit
+def _horn_slope(data, cellsize):
+    out = np.zeros_like(data)
+    rows, cols = data.shape
+    for y in range(1, rows-1):
+        for x in range(1, cols-1):
+            a = data[y+1, x-1]
+            b = data[y+1, x]
+            c = data[y+1, x+1]
+            d = data[y, x-1]
+            f = data[y, x+1]
+            g = data[y-1, x-1]
+            h = data[y-1, x]
+            i = data[y-1, x+1]
+            dz_dx = ((c + 2 * f + i) - (a + 2 * d + g)) / (8 * cellsize)
+            dz_dy = ((g + 2 * h + i) - (a + 2 * b + c)) / (8 * cellsize)
+            p = (dz_dx * dz_dx + dz_dy * dz_dy) ** .5
+            out[y, x] = np.arctan(p) * 57.29578
+    return out
+
+
+def slope(agg):
+    """Returns slope of input aggregate in degrees.
+
+    Parameters
+    ----------
+    agg : DataArray
+
+    Returns
+    -------
+    data: DataArray
+    """
+
+    if not isinstance(agg, DataArray):
+        raise TypeError("agg must be instance of DataArray")
+
+    if not agg.attrs.get('res'):
+        #TODO: maybe monkey-patch a "res" attribute valueing unity is reasonable
+        raise ValueError('input xarray must have numeric `res` attr.')
```

Comment:
Shouldn't the docstring explain that res is required, and what it should be?

Reply:
I've added the following note to the docstring:

```
    Notes:
    ------

    Input aggregate must contain a `res` (numeric) attribute which specifies the
    x/y cellsize for use in calculating slope.  It is assumed that cells are square
    and have the same resolution in x/y dimensions.

    Algorithm References:
     - http://desktop.arcgis.com/en/arcmap/10.3/tools/spatial-analyst-toolbox/how-slope-works.htm
     - Burrough, P. A., and McDonell, R. A., 1998. Principles of Geographical Information Systems (Oxford University Press, New York), pp 406
```